src/Umbraco.Web/PropertyEditors/MediaPicker3PropertyEditor.cs

```diff
@@ -22,43 +27,103 @@ namespace Umbraco.Web.PropertyEditors
+        /// Initializes a new instance of the <see cref="MediaPicker3PropertyEditor" /> class.
         /// </summary>
+        /// <param name="logger">The logger.</param>
         public MediaPicker3PropertyEditor(ILogger logger)
             : base(logger)
-        {
-        }
+        { }
 
         /// <inheritdoc />
         protected override IConfigurationEditor CreateConfigurationEditor() => new MediaPicker3ConfigurationEditor();
 
+        /// <inheritdoc />
         protected override IDataValueEditor CreateValueEditor() => new MediaPicker3PropertyValueEditor(Attribute);
 
         internal class MediaPicker3PropertyValueEditor : DataValueEditor, IDataValueReference
         {
-            ///<remarks>
-            /// Note: no FromEditor() and ToEditor() methods
-            /// We do not want to transform the way the data is stored in the DB and would like to keep a raw JSON string
-            /// </remarks>
-            public MediaPicker3PropertyValueEditor(DataEditorAttribute attribute) : base(attribute)
+            public MediaPicker3PropertyValueEditor(DataEditorAttribute attribute)
+                : base(attribute)
+            { }
+
+            public override object ToEditor(Property property, IDataTypeService dataTypeService, string culture = null, string segment = null)
             {
+                var value = property.GetValue(culture, segment);
+
+                return Deserialize(value);
+            }
+
+            public override object FromEditor(ContentPropertyData editorValue, object currentValue)
+            {
+                // TODO: Remove unneeded JSON properties (empty crops/default focal point) and remove indentation
+                return base.FromEditor(editorValue, currentValue);
             }
 
             public IEnumerable<UmbracoEntityReference> GetReferences(object value)
             {
-                var rawJson = value == null ? string.Empty : value is string str ? str : value.ToString();
+                foreach (var dto in Deserialize(value))
+                {
+                    yield return new UmbracoEntityReference(Udi.Create(Constants.UdiEntityType.Media, dto.MediaKey));
+                }
+            }
 
-                if (rawJson.IsNullOrWhiteSpace())
+            internal static IEnumerable<MediaWithCropsDto> Deserialize(object value)
+            {
+                var rawJson = value is string str ? str : value?.ToString();
+                if (string.IsNullOrWhiteSpace(rawJson))
+                {
                     yield break;
+                }
 
-                var mediaWithCropsDtos = JsonConvert.DeserializeObject<MediaPickerWithCropsValueConverter.MediaWithCropsDto[]>(rawJson);
-
-                foreach (var mediaWithCropsDto in mediaWithCropsDtos)
+                if (!rawJson.DetectIsJson())
+                {
+                    // Old comma seperated UDI format
+                    foreach (var udiStr in rawJson.Split(Constants.CharArrays.Comma))
```

Comment:
I still really would like to throw this in a try-catch in case people have weird values stored, logging the stored value will make it a lot easier to debug these cases which I am sure will come up :-) 

Maybe not a try-catch but at least if the `TryParse` fails we could log what weird value is in there. What do you think @ronaldbarendse ?

Reply:
There wasn't additional logging before and the Media Picker 2 even used `Udi.Parse` (without handling formatting/parsing errors) in the PVC:

https://github.com/umbraco/Umbraco-CMS/blob/d428a4543f33bb7094cf7db5f6b6fdc2d1de3063/src/Umbraco.Web/PropertyEditors/ValueConverters/MediaPickerValueConverter.cs#L59-L62

Because we can't do anything with incorrectly formatted UDIs, I think it's safe to simply ignore/skip them, just like the `GetReferences` implementation of MP2 does:

https://github.com/umbraco/Umbraco-CMS/blob/d428a4543f33bb7094cf7db5f6b6fdc2d1de3063/src/Umbraco.Web/PropertyEditors/MediaPickerPropertyEditor.cs#L48-L52

Reply:
@ronaldbarendse Really good point and in fact switching from MP3 back to MP2 already gave me ugly errors indeed, so it's safe to assume people wouldn't have any other data in there, it just would've failed to parse to begin with! 👍